From 2e7bd81e4097f9aa0c9a0583a2f392ed13b94d32 Mon Sep 17 00:00:00 2001
From: Anson Huang <Anson.Huang@nxp.com>
Date: Thu, 14 Jun 2018 14:41:23 +0800
Subject: [PATCH 4016/5242] MLK-18611-1 soc: imx: support MU off in suspend

commit  1f38a15fc3f685d2246ee63202fcc1e67ece42fe from
https://source.codeaurora.org/external/imx/linux-imx.git

Add MU restore function to support MU power off
in system suspend mode, need to re-initialize
MU after resume, since MU might lose power when
suspend.

Signed-off-by: Anson Huang <Anson.Huang@nxp.com>
Reviewed-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/sc/main/ipc.c |   16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/drivers/soc/imx/sc/main/ipc.c b/drivers/soc/imx/sc/main/ipc.c
index 05ae702..924f7a6 100644
--- a/drivers/soc/imx/sc/main/ipc.c
+++ b/drivers/soc/imx/sc/main/ipc.c
@@ -16,6 +16,7 @@
 #include <linux/interrupt.h>
 #include <linux/irq.h>
 #include <linux/mx8_mu.h>
+#include <linux/syscore_ops.h>
 
 #include <soc/imx8/sc/svc/irq/api.h>
 #include <soc/imx8/sc/ipc.h>
@@ -283,6 +284,19 @@ static irqreturn_t imx8_scu_mu_isr(int irq, void *param)
 	return IRQ_HANDLED;
 }
 
+static void imx8_mu_resume(void)
+{
+	int i;
+
+	MU_Init(mu_base_virtaddr);
+	for (i = 0; i < MU_RR_COUNT; i++)
+		MU_EnableGeneralInt(mu_base_virtaddr, i);
+}
+
+struct syscore_ops imx8_mu_syscore_ops = {
+	.resume = imx8_mu_resume,
+};
+
 /*Initialization of the MU code. */
 int __init imx8_mu_init(void)
 {
@@ -383,6 +397,8 @@ int __init imx8_mu_init(void)
 	if (sciErr)
 		pr_info("Cannot request WDOG interrupt\n");
 
+	register_syscore_ops(&imx8_mu_syscore_ops);
+
 	pr_info("*****Initialized MU\n");
 	return scu_mu_id;
 }
-- 
1.7.9.5

